File: addons/mail/models/res_role.py
import typing

from odoo import fields, models

if typing.TYPE_CHECKING:
    from odoo.addons.bus.models.res_users import ResUsers


class ResRole(models.Model):
    _name = "res.role"
    _description = (
        "Represents a role in the system used to categorize users. "
        "Each role has a unique name and can be associated with multiple users. "
        "Roles can be mentioned in messages to notify all associated users."
    )

    name = fields.Char(required=True)
    user_ids: ResUsers = fields.Many2many(
        comodel_name="res.users",
        relation="res_role_res_users_rel",
        string="Users",
    )

    _unique_name = models.UniqueIndex(
        "(name)", "A role with the same name already exists."
    )

File: addons/mail/static/tests/mock_server/mock_models/res_role.js
import { fields, models } from "@web/../tests/web_test_helpers";

export class ResRole extends models.ServerModel {
    _name = "res.role";

    name = fields.Char();
    color = fields.Char();
    sequence = fields.Integer({ default: 10 });
    user_ids = fields.Many2many({ relation: "res.users" });
    user_ids_count = fields.Integer({ compute: "_compute_user_ids_count" });

    _compute_user_ids_count() {
        for (const role of this) {
            role.user_ids_count = role.user_ids.length;
        }
    }
}
